### PR TITLE
Make build systems interchangeable (prepare for new alternative build system)

### PR DESCRIPTION
**Pull request by DropD**
## Description

Tweak the compiled backends code to allow every step of the process to be executed from the output of the previous step. Add Protocols for the steps as well as the step outputs to ensure that steps are as interchangeable as possible.

This will be the basis for two future directions of work:
* an additional build system, which uses CMake only once per session (and per language / dependency sets) to get a compile command and then skips CMake configure calls. (Use case: testing, interactive experimentation).
* caching and versioning of compilation results (by versioning I mean avoiding that different compiled versions of the same fencil clash).

## For reviewers

I would recommend starting with `gtfn_cpu`, as that change is what everything else builds up to.

## Requirements

Before submitting this PR, please make sure:

- [x] You have run the code checks, tests and documentation build successfully
- [x] All fixes and all new functionality are tested and documentation is up to date
- [x] You looked at the [review checklist](https://github.com/GridTools/gt4py/blob/master/REVIEW_CHECKLIST.txt)

Additionally, if this PR contains code authored by new contributors:

- [x] All the authors are covered by a valid contributor assignment agreement,
signed by the employer if needed, provided to ETH Zurich
- [x] The names of all the new contributors have been added to an updated
version of the AUTHORS.rst file included in the PR

### CODE REVIEW

File: src/functional/fencil_processors/builders/cpp/__init__.py
Comment (DropD) [inline src/functional/fencil_processors/builders/cpp/__init__.py:14]: Not needed anymore, since the `.callable` module is gone.

File: src/functional/fencil_processors/builders/cpp/bindings.py
Comment (DropD) [inline src/functional/fencil_processors/builders/cpp/bindings.py:174]: While it is true that this particular bindings creator does not require the full `source_module`, when caching and name mangling for versioning purposes come into play that might not be true anymore. In addition, we want to call each step directly with the output of the previous step.

This means, it might be better to combind this `language` with the `SourceModule.language` tag to further streamline the interface.

File: src/functional/fencil_processors/builders/cpp/build.py
Comment (DropD) [inline src/functional/fencil_processors/builders/cpp/build.py:139]: The thought behind making this a frozen dataclass is to make it hashable which could be beneficial for caching. It might be considered premature and comes at the cost that some of the properties (see below) have to be dynamically recomputed multiple times (although they are cheap).

There are two ways of solving this: either CMakeProject itself is again made stateful and the properties can be computed at construction (hashability is gone then) or the class is split into a dataclass which is hashable and a non-hashable class which can be created from it (this second class could be free of any hashing concerns and know only how to deal with files and the `cmake` command).

I am not sure it is worth deciding either way before we start implementing both the caching and the second build system. Both might influence the decision.

File: src/functional/fencil_processors/codegens/gtfn/gtfn_module.py
Comment (DropD) [inline src/functional/fencil_processors/codegens/gtfn/gtfn_module.py:33]: Have a look at `pipeline.py` for an explanation for the change to tuples.

File: tests/functional_tests/fencil_processor_tests/builder_tests/cpp_tests/test_cpp_pipeline.py
Comment (DropD) [inline tests/functional_tests/fencil_processor_tests/builder_tests/cpp_tests/test_cpp_pipeline.py:21]: This replaces both the `test_build` as well as `test_callable` tests.

File: src/functional/fencil_processors/source_modules/source_modules.py
Comment (DropD) [inline src/functional/fencil_processors/source_modules/source_modules.py:58]: Changing to tuple here and elsewhere to ensure they stay immutable (hashable), which is essential for caching to work well in the future.

File: src/functional/fencil_processors/runners/gtfn_cpu.py
Comment (DropD) [inline src/functional/fencil_processors/runners/gtfn_cpu.py:61]: This is the pipeline in action. Every step feeds directly into the next. Most of the other changes in this PR are aimed at achieving this.

File: src/functional/fencil_processors/runners/gtfn_cpu.py
Comment (DropD) [inline src/functional/fencil_processors/runners/gtfn_cpu.py:65]: This one is not strictly necessary but allows for easier to read test fixture ids.

File: src/functional/fencil_processors/runners/gtfn_cpu.py
Comment (DropD) [inline src/functional/fencil_processors/runners/gtfn_cpu.py:39]: Why turn this into a class? This comes from a discussion with @petiaccja a while back, where we thought of customizing fencil processors by generating them with customization, rather than passing configuration through the keyword args.

In this case I chose a class, rather than a function generating a function because I suspect it is more readable to the uninitiated. Also it allows naming the configured processor, which helps with test fixture ids.

File: src/functional/fencil_processors/codegens/gtfn/gtfn_module.py
Comment (DropD) [inline src/functional/fencil_processors/codegens/gtfn/gtfn_module.py:45]: This has become a class for the same reason as `gtfn_cpu.GTFNExecutor`. More detail there.

File: src/functional/fencil_processors/source_modules/cpp_gen.py
Comment (petiaccja) [inline src/functional/fencil_processors/source_modules/cpp_gen.py:31]: If we identify languages by the class type, `name` here is redundant. If we still want a string name for messages to the user, it might be nicer to use pretty names like `"C++"` and `"Python"`.

File: src/functional/fencil_processors/source_modules/cpp_gen.py
Comment (petiaccja) [inline src/functional/fencil_processors/source_modules/cpp_gen.py:34]: Maybe a comment on where to find the available options for formatting styles? Perhaps an enum in eve?

File: src/functional/fencil_processors/source_modules/source_modules.py
Comment (petiaccja) [inline src/functional/fencil_processors/source_modules/source_modules.py:34]: How about `LanguageFileExtensionsProtocol` or `SourceFileExtensionsProtocol` or something similar? I interpreted the name as "include the ImplementationLanguageProtocol" and it was pretty confusing.

**Review by petiaccja — CHANGES_REQUESTED** [review]

File: src/functional/fencil_processors/source_modules/cpp_gen.py
Comment (DropD) [reply src/functional/fencil_processors/source_modules/cpp_gen.py:43]: I see it like this: if someone writes a backend that separates headers cleanly, then they are free to provide a different default `CppLanguage` instance, which sets `implementation_extension` to "hpp".

Only once someone really writes a backend that needs three estensions at the same time should there be a new language protocol and a default implementation for it. Seems pointless otherwise.

File: src/functional/fencil_processors/source_modules/cpp_gen.py
Comment (DropD) [reply src/functional/fencil_processors/source_modules/cpp_gen.py:42]: Good point.

File: src/functional/fencil_processors/source_modules/cpp_gen.py
Comment (DropD) [reply src/functional/fencil_processors/source_modules/cpp_gen.py:34]: I will leave this up to eve...

File: src/functional/fencil_processors/source_modules/source_modules.py
Comment (DropD) [reply src/functional/fencil_processors/source_modules/source_modules.py:34]: * `LanguageFileExtensionsProtocol`: looks like `Python` might want to implement that. Except python does not split "include" and "implementation" files. -> Too general
* `SourceFileExtensionsProtocol`: `Python` also has source file extensions...
* `LanguagesWhichSplitImplementationAndIncludeFilesProtocol`: too long
* `HeaderAndImplementationFilesLanguageProtocol`: a bit long.

Considering we don't really need the `Protocol`, maybe `LanguageWithHeaders`?

File: src/functional/fencil_processors/source_modules/source_modules.py
Comment (egparedes) [inline src/functional/fencil_processors/source_modules/source_modules.py:83]: If `SourceModule` is generic with respect to a `LanguageT` typevar, shouldn't `BindingModule` be generic with respect to the target language, and maybe even also with respect to the `SourceModule` language (contravariantly)?

File: src/functional/fencil_processors/pipeline.py
Comment (egparedes) [inline src/functional/fencil_processors/pipeline.py:24]: Just a question: why is the type not covariantly parametrized with respect to the output binding language?

File: src/functional/fencil_processors/builders/cpp/build.py
Comment (egparedes) [inline src/functional/fencil_processors/builders/cpp/build.py:30]: Too many direct imports: https://google.github.io/styleguide/pyguide.html#22-imports

**Review by egparedes — COMMENTED** [review]
I have several comments but it looks good. The main issue in my opinion is the design of the Language Protocol as a protocol for instances and not for classes used as singleton/enums. Check my comments for further details and discussion.

File: src/functional/fencil_processors/pipeline.py
Comment (DropD) [reply src/functional/fencil_processors/pipeline.py:24]: This took me a while to get (mypy complained about using the covariant type var), because I hadn't understood co- and contravariant all the way.

A BindingsGenerator written for the most general language protocol works with specific ones but not vice-versa. So you can substitute *super*-classes for *sub*-classes (contravariant) and not *sub*-classes for *super*-classes (covariant).

File: src/functional/fencil_processors/source_modules/source_modules.py
Comment (DropD) [reply src/functional/fencil_processors/source_modules/source_modules.py:83]: Probably should be 👍

File: src/functional/fencil_processors/pipeline.py
Comment (DropD) [reply src/functional/fencil_processors/pipeline.py:24]: On the other hand, now mypy complains the other way around, so I apparently don't understand anything.

File: src/functional/fencil_processors/runners/gtfn_cpu.py
Comment (DropD) [reply src/functional/fencil_processors/runners/gtfn_cpu.py]: It is this function which is missing something, namely support for arguments other than scalars and located fields (for example constant fields are not supported). That support is explicitly out of scope, however if the tests pass without wrapping arrays in `memoryview`, then I don't mind removing that.

File: src/functional/fencil_processors/pipeline.py
Comment (egparedes) [reply src/functional/fencil_processors/pipeline.py]: In my experience, understanding covariance and contravariance is much simpler than understanding what the hell is mypy understanding for covariance and contravariance. I almost gave up completely on using covariant and contravariant typevars in Python annotations...

File: src/functional/fencil_processors/runners/gtfn_cpu.py
Comment (DropD) [reply src/functional/fencil_processors/runners/gtfn_cpu.py]: Funny story: I commented out the `memoryview` line and suddenly all the builtins tests passed locally for the first time for me. Could have been something different but the change didn't make anything fail, so I am keeping it.

File: src/functional/fencil_processors/runners/gtfn_cpu.py
Comment (DropD) [reply src/functional/fencil_processors/runners/gtfn_cpu.py]: My intention was actually not to support mypy in deducing the type. I am expressing that this should always implement the protocol. Is there a better way to warn future developers immediately in case of accidental incompatible changes?

File: src/functional/fencil_processors/source_modules/source_modules.py
Comment (DropD) [reply src/functional/fencil_processors/source_modules/source_modules.py:28]: It is intended to label different sets of language preferences (instances). If it was a class var it would be redundant (The class already has a readable name) and you would only be able to tell one set of `C++` preferences from another by looking at each of the values.

File: src/functional/fencil_processors/source_modules/source_modules.py
Comment (DropD) [reply src/functional/fencil_processors/source_modules/source_modules.py:43]: Same reply as above. If someone needs all the headers to be called `.my.custom.hpp` for their integration purposes, they would need to implement a new (otherwise redundant) class.

File: src/functional/fencil_processors/source_modules/cpp_gen.py
Comment (DropD) [reply src/functional/fencil_processors/source_modules/cpp_gen.py:28]: Protocols are in a way necessarily singletons (only the class exists, no instances). However, the implementations are explicitly designed to not be singletons. There is a usable default value, which is not instanciated everywhere separately, which may look like a singleton?

**Review by petiaccja — CHANGES_REQUESTED** [review]

File: src/functional/fencil_processors/source_modules/source_modules.py
Comment (petiaccja) [reply src/functional/fencil_processors/source_modules/source_modules.py]: A bit awkward still, but I can live with that.

File: src/functional/fencil_processors/runners/gtfn_cpu.py
Comment (petiaccja) [reply src/functional/fencil_processors/runners/gtfn_cpu.py]: Yeah, this is a mess. 

One problem is that numpy will convert literally anything to an array with `asarray`, even scalars. You must check the dimensions afterwards.

As far as I know, the buffer protocol is supported if the object can be converted to a memory view. The other problem is that our `Field` class does not support the buffer protocol, but it does support conversion to numpy arrays, and numpy arrays do support the buffer protocol.

The real fix to this would be to make `Field` support the buffer protocol and do an attempt to cast the argument into a memory view to detect support.

File: src/functional/fencil_processors/builders/cpp/bindings.py
Comment (DropD) [reply src/functional/fencil_processors/builders/cpp/bindings.py]: This is outdated indeed. It was there to isolate the mypy error to the `memoryview` line which was removed later.

File: src/functional/fencil_processors/builders/cpp/bindings.py
Comment (DropD) [reply src/functional/fencil_processors/builders/cpp/bindings.py]: 👍

File: src/functional/fencil_processors/pipeline.py
Comment (DropD) [reply src/functional/fencil_processors/pipeline.py:32]: Definitely an improvement!

File: src/functional/fencil_processors/runners/gtfn_cpu.py
Comment (DropD) [reply src/functional/fencil_processors/runners/gtfn_cpu.py:55]: I would like to avoid that. It is an invitation to put custom logic between the steps (rather than improve the steps). If this is not readable enough, we indeed need a pipeline runner of some kind. Something along the lines of

```python
return <pipeline_runner_name>(
    source_module_gen=gtfn_module.GTFNSourceModuleGenerator(self.language_settings),
    binding_module_gen=bindings.create_bindings,
    buildable_project_gen=build.CMakeProject,
    cache_strategy=cache.Strategy.SESSION,
)(fencil).get_implementation()
```

Comment (DropD) [reply src/functional/fencil_processors/runners/gtfn_cpu.py:55]: I assume you meant something like

```python
# Never ever put any logic in between these lines!
# Instead, if something is missing it has to go into one of the functions we are calling here.
source_module = gtfn_module.GTFNSourceModuleGenerator(self.language_settings)(fencil, *args, **kwargs)
bindings_module = bindings.create_bindings(source_module)
cmake_project =  build.CMakeProject(
    source_module=source_module,
    bindings_module=bindings_module,
    cache_strategy=cache.Strategy.SESSION,
)
cpp_python_extension_function = cmake_project.get_implementation()
cpp_python_extension_function(*[convert_arg(arg) for arg in args])
```

File: src/functional/fencil_processors/source_modules/cpp_gen.py
Comment (DropD) [reply src/functional/fencil_processors/source_modules/cpp_gen.py:34]: I put that into the docstring of `source_modules.LanguageSettings`

File: src/functional/fencil_processors/builders/cpp/build.py
Comment (DropD) [reply src/functional/fencil_processors/builders/cpp/build.py:138]: I would like to do that in the following PR, where I am touching the `CMakeProject` anyway. Also to avoid conflicts.

Comment (petiaccja) [reply src/functional/fencil_processors/builders/cpp/build.py:138]: If you think this form is better for this PR and you remove it in the next, it's fine for me.

File: src/functional/fencil_processors/source_modules/cpp_gen.py
Comment (petiaccja) [reply src/functional/fencil_processors/source_modules/cpp_gen.py]: What use case have you had in mind for this extensions feature? I don't really understand how it's connected to unique names.

Include guards don't help here, they prevent multiple inclusions into the same translation unit, but they allow inclusion into many translation units. The problem with `.cpp.inc` is that it should be included into no more than one translation unit.

File: src/functional/fencil_processors/pipeline.py
Comment (DropD) [reply src/functional/fencil_processors/pipeline.py]: side note: what mypy expects depends on the presence (or not) of the infamous `src/__init__.py`. So this might need to be revisited...

**Review by egparedes — COMMENTED** [review]
It looks good to me. Just minor comments and questions.

File: src/functional/fencil_processors/runners/gtfn_cpu.py
Comment (petiaccja) [reply src/functional/fencil_processors/runners/gtfn_cpu.py:55]: Yeah, let's leave it like this with a TODO

File: src/functional/fencil_processors/builders/cpp/build.py
Comment (petiaccja) [reply src/functional/fencil_processors/builders/cpp/build.py:138]: Rico, can you put a TODO on this before merging?

File: src/functional/fencil_processors/runners/gtfn_cpu.py
Comment (petiaccja) [reply src/functional/fencil_processors/runners/gtfn_cpu.py]: How about leaving the mess and waiting for the no-storage implementation? That'll make the problem go away, right?

Comment (egparedes) [reply src/functional/fencil_processors/runners/gtfn_cpu.py]: Not at all. The no-storage implementation in functional will only add the feature of allocating buffers with specific memory layouts.

Comment (egparedes) [reply src/functional/fencil_processors/runners/gtfn_cpu.py]: Hmmm, it's not exactly correct what I said before, let me rephrase it: the no-storage implementation will not solve this problem in particular but it will require to deal with field buffers in a different and more coherent way.

Comment (egparedes) [reply src/functional/fencil_processors/runners/gtfn_cpu.py]: Anyway, the important point of my suggestion is that we should restrict the conversion to ndarray only when the argument defines an `__array__` method, not always.

File: src/functional/fencil_processors/source_modules/cpp_gen.py
Comment (DropD) [reply src/functional/fencil_processors/source_modules/cpp_gen.py:34]: Nope, dead code.

File: src/functional/fencil_processors/source_modules/cpp_gen.py
Comment (DropD) [reply src/functional/fencil_processors/source_modules/cpp_gen.py:37]: It's dead, thanks.

File: src/functional/fencil_processors/codegens/gtfn/gtfn_module.py
Comment (DropD) [reply src/functional/fencil_processors/codegens/gtfn/gtfn_module.py]: I changed it in both places, thanks for noticing.

File: src/functional/fencil_processors/runners/gtfn_cpu.py
Comment (egparedes) [inline src/functional/fencil_processors/runners/gtfn_cpu.py:35]: Not critical for this PR, but still suggest to add a ToDo comment here or change it to something more restrictive like:
```python
if hasattr(arg, "__array__"):
    return np.asarray(arg)
else:
    return arg

```

**Review by egparedes — APPROVED** [review]
It looks good.

File: src/functional/fencil_processors/builders/cpp/bindings.py
Comment (DropD) [reply src/functional/fencil_processors/builders/cpp/bindings.py]: The problem was actually in `test_bindings.py`. `ScalarParameter.dtype` is not validated, as opposed to `FunctionParameter.dtype`, so the test there was able to sneak in a `_ctypes.PyCSimpleType` instead.

File: src/functional/fencil_processors/runners/gtfn_cpu.py
Comment (DropD) [reply src/functional/fencil_processors/runners/gtfn_cpu.py]: I just went ahead and changed it, if it breaks tests I will revert and add a TODO.

**Review by petiaccja — APPROVED** [review]

**Pull request merged**
